clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsServerClusterProvider.java

```diff
@@ -0,0 +1,269 @@
+      if (credentials instanceof AmazonCredentials) {
+        clusterMap = findClusters(clusterMap, (AmazonCredentials) credentials);
+      }
+    }
+    return clusterMap;
+  }
+
+  private Map<String, Set<EcsServerCluster>> findClusters(Map<String, Set<EcsServerCluster>> clusterMap,
+                                                          AmazonCredentials credentials) {
+    for (AmazonCredentials.AWSRegion awsRegion: credentials.getRegions()) {
+      clusterMap = findClustersForRegion(clusterMap, credentials, awsRegion);
+    }
+
+    return clusterMap;
+  }
+
+  private Map<String, Set<EcsServerCluster>> findClustersForRegion(Map<String, Set<EcsServerCluster>> clusterMap,
+                                                                   AmazonCredentials credentials,
+                                                                   AmazonCredentials.AWSRegion awsRegion) {
+
+    AmazonECS amazonECS = amazonClientProvider.getAmazonEcs(credentials.getName(), credentials.getCredentialsProvider(), awsRegion.getName());
+    AmazonEC2 amazonEC2 = amazonClientProvider.getAmazonEC2(credentials.getName(), credentials.getCredentialsProvider(), awsRegion.getName());
+    AmazonElasticLoadBalancing amazonELB = amazonClientProvider.getAmazonElasticLoadBalancingV2(credentials.getName(), credentials.getCredentialsProvider(), awsRegion.getName());
+
+    for (String clusterArn: amazonECS.listClusters().getClusterArns()) {
+      String ecsClusterName = inferClusterNameFromClusterArn(clusterArn);
+
+      ListServicesResult result = amazonECS.listServices(new ListServicesRequest().withCluster(ecsClusterName));
+      for (String serviceArn: result.getServiceArns()) {
+
+        ServiceMetadata metadata = extractMetadataFromServiceArn(serviceArn);
+        Set<Instance> instances = new HashSet<>();
+
+        DescribeLoadBalancersResult loadBalancersResult = amazonELB.describeLoadBalancers(new DescribeLoadBalancersRequest());
+        Set<com.netflix.spinnaker.clouddriver.model.LoadBalancer> loadBalancers = extractLoadBalancersData(loadBalancersResult);
+
+        ListTasksResult listTasksResult = amazonECS.listTasks(new ListTasksRequest().withServiceName(serviceArn).withCluster(ecsClusterName));
+        if (listTasksResult.getTaskArns() != null && listTasksResult.getTaskArns().size() > 0) {
+          DescribeTasksResult describeTasksResult = amazonECS.describeTasks(new DescribeTasksRequest().withCluster(ecsClusterName).withTasks(listTasksResult.getTaskArns()));
+
+          for (Task task: describeTasksResult.getTasks()) {
+            InstanceStatus ec2InstanceStatus = containerInformationService.getEC2InstanceStatus(amazonEC2, containerInformationService.getContainerInstance(amazonECS, task));
+            instances.add(new EcsTask(extractTaskIdFromTaskArn(task.getTaskArn()), task, ec2InstanceStatus));
+          }
+        }
+
+        EcsServerGroup ecsServerGroup = generateServerGroup(awsRegion, metadata, instances);
```

Comment:
Nitpicking, generateServerGroup and generateSpinnakerServerCluster seem unnecessary, all they do is call the constructor and return the object.

Reply:
They make the code more readable by allowing it to tell a story with human words and communicate intentionality, rather than out-of-context code.  Especially likable in a long method that really is too long to respect the "do one thing but do it well" principle